Build aarch64-darwin binaries

ch1bo opened the pull request:
:apple: Adds a workflow to build binaries on custom MacOS with Apple Silicon runner (`aarch64-darwin` system)

:apple: ~~Packages binary with runtime dependencies into a single archive so users can use `DYLD_FALLBACK_LIBRARY_PATH` to run the binaries (without further installing)~~

:apple: Use static linking for third-party libraries on all darwin builds and do some patching (courtesy of @angerman) when packing them up (in a `nix build .#release)

Example builds:
* Linux x86_64: https://github.com/input-output-hk/hydra/actions/runs/7037060630
* MacOS aarch64: https://github.com/input-output-hk/hydra/actions/runs/7037060637

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

abailly committed:
Configure job for darwin

abailly committed:
Dedicated build-only job for Darwin

abailly committed:
Rename build step

ch1bo committed:
Add some documentation to workflows

ch1bo committed:
Collect runtime libraries from nix store

Using ldd we can find and copy the runtime dependencies and later use
DYLD_FALLBACK_LIBRARY_PATH to point there.

ch1bo committed:
Use otool instead

abailly committed:
Minor typo

ghost commented:
Seems like the produced binaries are not runnable, I have successfully retrieved the produced package but the binaries crash even when given the `DYLD_FALLBACK_LIBRARY_PATH` path to find their libraries.

ch1bo committed:
Modify packaging.nix to produce our zip format

We do not want a directory inside and it was not packaging correctly
multiple derivations.

ch1bo committed:
Update workflows to use .#release and .#release-static

ch1bo committed:
Use mkDerivation instaed of runCommand in embedRevision

This ensures the resulting derivation has the same stdenv and downstream
functions can infer the targetPlatform correctly, e.g. packaging.nix

ch1bo committed:
Use artifact names based on branch/tag

ch1bo committed:
Create versioned artifacts

ch1bo committed:
Renamed workflows

Tried to use a somewhat consistent and not confusing (not too technical)
naming. That is, no system triplets and something users would understand
on the top level.

ch1bo committed:
Fix github artifacts

Upload artifact is always zipping the paths provided, so we need to
unpack the release archive first.

ch1bo committed:
Avoid .zip.zip in artifact download

ch1bo commented:
> Seems like the archive is published with twice the `.zip` extension
> 
> ![Screenshot 2023-11-30 at 11 06 21](https://private-user-images.githubusercontent.com/79840582/286892554-20863afc-8c33-4679-bf7b-547d0e48a4da.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE0MjY4MDEsIm5iZiI6MTcwMTQyNjUwMSwicGF0aCI6Ii83OTg0MDU4Mi8yODY4OTI1NTQtMjA4NjNhZmMtOGMzMy00Njc5LWJmN2ItNTQ3ZDBlNDhhNGRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAxVDEwMjgyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3ZTRiNTFmY2M5NDQwY2EwNDhmZmFlM2Y2NzlkMzI4ZmNjNDA0NjAwN2IwYTZkM2Y2NmQ2ZDJkOGU5MDM3ZTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pv0JXH_Oze3VdkN7iSclvT9OTFVpYA_mKhJc0DfFM7s)

Was only double the extension. Fixed now https://github.com/input-output-hk/hydra/pull/1182/commits/e14bc41a528dc732c4af914e575297d29c20d92f

ghost commented:
I guess the hydra errors are spurious? We have observed those already in the past IIRC

v0d1ch commented:
Are we ok with darwin builds failing here? Do we want to support it going forward or we are hopping onto silicon train? :)
